pkg/controller/syncsetinstance/syncsetinstance_controller.go

```diff
@@ -574,119 +638,130 @@ func (r *ReconcileSyncSetInstance) applySyncSetPatches(ssi *hivev1.SyncSetInstan
+			secretReferenceSyncStatus.Conditions = r.setApplySyncConditions(secretReferenceSyncConditions, readErr)
+			syncStatusList = append(syncStatusList, secretReferenceSyncStatus)
+			break
 		}
-		gvr := gv.WithResource(deletedStatus.Resource)
-		itemLog.Debug("deleting resource")
-		err = dynamicClient.Resource(gvr).Namespace(deletedStatus.Namespace).Delete(deletedStatus.Name, &metav1.DeleteOptions{})
-		if err != nil {
-			if !errors.IsNotFound(err) {
-				itemLog.WithError(err).Error("error deleting resource")
-				index := deletedStatusIndices[i]
-				existingStatusList[index].Conditions = r.setDeletionFailedSyncCondition(existingStatusList[index].Conditions, err)
+
+		secret.Name = secretReference.Target.Name
+		secret.Namespace = secretReference.Target.Namespace
+		// These pieces of metadata need to be set to nil values to perform an update from the original secret
+		secret.Generation = 0
+		secret.ResourceVersion = ""
+		secret.UID = ""
+
+		var hash string
+		hash, hashErr = controllerutils.GetChecksumOfObject(secret)
+		if hashErr != nil {
+			ssiLog.WithError(hashErr).WithField("secret", fmt.Sprintf("%s/%s", secretReference.Source.Name, secretReference.Source.Namespace)).Error("unable to compute secret hash")
+			secretReferenceSyncStatus.Conditions = r.setApplySyncConditions(secretReferenceSyncConditions, hashErr)
+			syncStatusList = append(syncStatusList, secretReferenceSyncStatus)
+			break
+		}
+		secretReferenceSyncStatus.Hash = hash
+
+		if rss == nil || needToReApply("secret", secretReferenceSyncStatus, *rss, ssiLog) {
+			// Apply secret
+			ssiLog.Debugf("applying secret: %s/%s (%s)", secret.Namespace, secret.Name, secret.Kind)
+			var result hiveresource.ApplyResult
+			result, applyErr = h.ApplyRuntimeObject(secret, scheme.Scheme)
+
+			var secretReferenceSyncConditions []hivev1.SyncCondition
+			if rss != nil {
+				secretReferenceSyncConditions = rss.Conditions
+			}
+			secretReferenceSyncStatus.Conditions = r.setApplySyncConditions(secretReferenceSyncConditions, applyErr)
+
+			if applyErr != nil {
+				ssiLog.WithError(applyErr).Errorf("error applying secret %s/%s (%s)", secret.Namespace, secret.Name, secret.Kind)
 			} else {
-				itemLog.Debug("resource not found, nothing to do")
+				ssiLog.Debugf("resource %s/%s (%s): %s", secret.Namespace, secret.Name, secret.Kind, result)
 			}
+		} else {
+			// Do not apply secret
+			ssiLog.Debugf("resource %s/%s (%s) has not changed, will not apply", secret.Namespace, secret.Name, secret.Kind)
+			secretReferenceSyncStatus.Conditions = rss.Conditions
+		}
+
+		syncStatusList = append(syncStatusList, secretReferenceSyncStatus)
+
+		// If an error applying occurred, stop processing right here
+		if applyErr != nil {
+			break
 		}
 	}
 
-	return newStatusList, nil
+	ssi.Status.SecretReferences = r.reconcileDeleted("secret", ssi.Spec.ResourceApplyMode, dynamicClient, ssi.Status.SecretReferences, syncStatusList, applyErr, ssiLog)
```

Comment:
The `err` parameter passed to `reconcileDeleted` should be non-nil if any of `readErr`, `hashErr`, or `applyErr` are non-nil. I don't know if there is any reason to keep the three classes of errors separate. Ultimately, each one is just an error applying the secret to the remote cluster, whether that is an error in getting the data to apply or in sending the data to apply.

Reply:
Updated to only use `applyErr` for each case.